src/server/pfs/persist/driver.go

```diff
@@ -0,0 +1,462 @@
+		return err
+	}
+
+	_, err = d.getTerm(repoTable).Insert(&Repo{
+		Name:    repo.Name,
+		Created: created,
+	}).RunWrite(d.dbClient)
+	return err
+}
+
+func (d *driver) InspectRepo(repo *pfs.Repo, shards map[uint64]bool) (repoInfo *pfs.RepoInfo, retErr error) {
+	cursor, err := d.getTerm(repoTable).Get(repo.Name).Default(gorethink.Error("value not found")).Run(d.dbClient)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := cursor.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	rawRepo := &Repo{}
+	cursor.Next(rawRepo)
+	if err := cursor.Err(); err != nil {
+		return nil, err
+	}
+	repoInfo = &pfs.RepoInfo{
+		Repo:    &pfs.Repo{rawRepo.Name},
+		Created: rawRepo.Created,
+	}
+	return repoInfo, nil
+}
+
+func (d *driver) ListRepo(provenance []*pfs.Repo, shards map[uint64]bool) (repoInfos []*pfs.RepoInfo, retErr error) {
+	cursor, err := d.getTerm(repoTable).Run(d.dbClient)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := cursor.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	for {
+		repo := &Repo{}
+		if !cursor.Next(repo) {
+			break
+		}
+		repoInfos = append(repoInfos, &pfs.RepoInfo{
+			Repo:    &pfs.Repo{repo.Name},
+			Created: repo.Created,
+		})
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, err
+	}
+	return repoInfos, nil
+}
+
+func (d *driver) DeleteRepo(repo *pfs.Repo, shards map[uint64]bool, force bool) error {
+	_, err := d.getTerm(repoTable).Get(repo.Name).Delete().RunWrite(d.dbClient)
+	return err
+}
+
+func (d *driver) StartCommit(repo *pfs.Repo, commitID string, parentID string, branch string, started *google_protobuf.Timestamp, provenance []*pfs.Commit, shards map[uint64]bool) (retErr error) {
+	if commitID == "" {
+		commitID = uuid.NewWithoutDashes()
+	}
+
+	var _provenance []string
+	for _, c := range provenance {
+		_provenance = append(_provenance, c.ID)
+	}
+	commit := &Commit{
+		ID:         commitID,
+		Repo:       repo.Name,
+		Started:    prototime.TimeToTimestamp(time.Now()),
+		Provenance: _provenance,
+	}
+
+	if parentID == "" {
+		if branch == "" {
+			branch = uuid.NewWithoutDashes()
+		}
+
+		_, err := d.getTerm(branchTable).Insert(&Branch{
+			ID: branchID(repo.Name, branch),
+		}, gorethink.InsertOpts{
+			// Set conflict to "replace" because we don't want to get an error
+			// when the branch already exists.
+			Conflict: "replace",
+		}).RunWrite(d.dbClient)
+		if err != nil {
+			return err
+		}
+
+		cursor, err := d.getTerm(commitTable).OrderBy(gorethink.OrderByOpts{
+			Index: gorethink.Desc(commitBranchIndex),
+		}).Between(
+			[]interface{}{branch, 0},
+			[]interface{}{branch, gorethink.MaxVal},
+		).Run(d.dbClient)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := cursor.Close(); err != nil && retErr == nil {
+				retErr = err
+			}
+		}()
+
+		// The last commit on this branch will be our parent commit
+		parentCommit := &Commit{}
+		found := cursor.Next(parentCommit)
```

Comment:
Isn't this potentially racy if multiple nodes are trying to start commits on the same branch?